manifests/kerberos.pp

```diff
@@ -115,5 +132,86 @@
+      # Step 1: Ensure a temporary script file is in place only if a command fails
+      exec { 'check_keytab_command':
+        path      => ['/usr/bin', '/usr/sbin', '/usr/lib/mit/bin'],
+        command   => "echo 'host is missing from keytab' && touch /root/ad_createhostkeytab.sh && chmod 0700 /root/ad_createhostkeytab.sh",
+        unless    => "klist -kte 2>&1 | grep 'host/${facts['networking']['fqdn']}' | grep -i '${ad_domain}'",
+        logoutput => true,
+        notify    => [
+          Exec['create_ad_createhostkeytab.sh_script'],
+          Exec['run_ad_create_host_keytab_script'],
+        ],
+      }
+
+      # File RESOURCE CANNOT ONLY RUN WHEN EXEC ABOVE RUNS
+      # SO USING Exec RESOURCE TO CREATE THE SCRIPT FILE
+      exec { 'create_ad_createhostkeytab.sh_script':
+        path        => ['/usr/bin', '/usr/sbin'],
+        command     => "cat <<'EOF' > /root/ad_createhostkeytab.sh
+#!/bin/bash
+# ASSIGN ARGUMENTS FROM COMMAND LINE ARGUMENTS
+ad_domain=\"\$1\"   # Active Directory domain
+ad_ou_computers=\"\$2\"  # Organizational Unit for computers in AD
+ad_user=\"\$3\"  # User with permissions to create host in AD
+keytab_base64=\"\$4\"  # Base64 encoded keytab for creating host
+# ASSIGN STATIC VARIABLES
+HOST_FQDN=\"\$(hostname -f)\" # Fully Qualified Domain Name of the host
+keytab_file=\"/root/createhost.keytab\"  # Path to store the decoded keytab file
+os_name=\"\$(facter os.name)\"  # Operating System name
+# Decode the base64 encoded keytab and save it to a file
+echo \"\${keytab_base64}\" | base64 --decode > \${keytab_file}
+# Authenticate using the keytab file
+kinit -k -t \$keytab_file \$ad_user
+# Pre-create the computer account in the specified OU
+adcli preset-computer --domain=\"\${ad_domain}\" --domain-ou=\"\${ad_ou_computers}\" \
+  -U \"\${ad_user}\" --login-ccache --os-name=\"\${os_name}\" \$HOST_FQDN
+# Join the computer to the AD domain
+adcli join --domain=\"\${ad_domain}\" -U \"\${ad_user}\" --login-ccache
+# Destroy the Kerberos ticket cache for the user
+kdestroy -p \$ad_user
+# Remove the keytab file for security reasons
+rm -f \$keytab_file
+EOF
+",
+        refreshonly => true,
+        require     => [
+          Exec['check_keytab_command'],
+        ],
+      }
+
+      # Step 2: Run the script if the command fails
+      exec { 'run_ad_create_host_keytab_script':
+        path        => ['/usr/bin', '/usr/sbin', '/usr/lib/mit/bin'],
+        command     => Sensitive(
+          "/root/ad_createhostkeytab.sh '${ad_domain}' '${ad_computers_ou}' '${ad_createhostuser}' '${ad_createhostkeytab}' "
+        ),
+        refreshonly => true,
+        require     => [
+          Exec['check_keytab_command'],
+          Exec['create_ad_createhostkeytab.sh_script'],
+        ],
+        notify      => Exec['delete_ad_create_host_keytab_script'],
+      }
+
+      # Step 3: Ensure the temporary script file is deleted once it has been run
+      exec { 'delete_ad_create_host_keytab_script':
```

Comment:
Why delete the /root/ad_createhostkeytab.sh script? Unless I'm missing something it looks like all the sensitive data is in the arguments for the script and not in the script itself.

Reply:
Yes, that was a late pivot to remove all the sensitive data from the script. I wrote the logic for removing the script before I ended up removing the sensitive data from the script.

It could be left on the host, but there really isn't much use for it to stay there. But the removal of the script was inspired from the current #19 issue where it was requested that the older `createhostkeytab.sh` be removed.

Are suggesting that there is a reason to keep the script around?

Reply:
Gotcha, no real reason to keep it, but knowing there isn't anything sensitive in the script I think the logic can be simplified.

Make a custom fact with the command `"klist -kte 2>&1 | grep 'host/${facts['networking']['fqdn']}' | grep -i '${ad_domain}'"` to be something like `is_joined_to_ad_domain`

Since that is now a fact and there is no sensitive info in the script itself, you should be safe to go back to using the File resource and the template to create the /root/ad_createhostkeytab.sh script.

So something like this:

```
if ( $ad_createhostkeytab and $ad_createhostuser and $ad_computers_ou and $ad_domain ) {

   if $facts['is_joined_ad_domain'] {
      $ensure_parm = 'absent'
   } else {
      $ensure_parm = 'present'

      exec { 'run_ad_create_host_keytab_script':
         path        => ['/usr/bin', '/usr/sbin', '/usr/lib/mit/bin'],
         command     => Sensitive(
           "/root/ad_createhostkeytab.sh '${ad_domain}' '${ad_computers_ou}' '${ad_createhostuser}' '${ad_createhostkeytab}' "
        ),
        refreshonly => true,
        require => File['/root/ad_createhostkeytab.sh'],        
      }

   }

   file { '/root/ad_createhostkeytab.sh':
      ensure    => $ensure_parm,
      content => #grab content from your template
      ... rest of standard stuff here ...
   }

}
```

Reply:
Thank you!!!  That's great advice!!